### PR TITLE
[DO NOT MERGE] SRP POC

### PR DESCRIPTION
**Pull request by FRSTR**

### CODE REVIEW

**Comment by FRSTR** [comment]
This new spr.py file only needs renault-api installed in the system (I tested it in WSL), so it does not depend on other files in this repository.
At the top of the files there are 4 vars with user IDs, vin and pincode to fill.
Sometimes I saw this flow fails earlier, but in general 90% of the time it should try to complete all the steps. The final step does shoudl show that the command was CANCELLED (because srp auth fails now) - there is no logic to check this.

File: spr.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    print("========================")
+    print("1. RES-STATE")
+    print("========================")
+    strExec = (
+        'renault-api --debug http get "/commerce/v1/accounts/'
+        + AccountId
+        + "/kamereon/kca/car-adapter/v1/cars/"
+        + vin
+        + '/res-state?country=RU"'
+    )
```
[the remark this answers is not in the document]
Comment (epenet) [reply spr.py:44]: Note: I think this would be useful to add to HA, independantly of the SRP methods

[the remark this answers is not in the document]
Comment (epenet) [reply spr.py]: It seems to work like this:

1. GET `res-state` => checks the state of the remote engine. I wonder what other codes are available.
2. POST `actions-srp-sets` => starts authentication
3. GET `notifications` => wait for authentication confirmation (actionType = COMMAND_RESPONSE)
4. GET `notifications` => wait for authentication confirmation (actionType = SRP_SALT_REQUEST)
5. POST `actions-engine-start` => send the actual start-engine command

File: spr.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    stream = os.popen(strExec)
+    # output = stream.read()
+    time.sleep(5)
+
+    output = " ".join(stream)
+    output = output.replace('"', "")
+    output = output.replace("'", "")
+    output = output.replace(" ", "")
+
+    ParseList1 = output.split("SrpSets,id:")
```
Comment (epenet) [inline spr.py:82]: I think this is very difficult to understand.
You should use JSON parsing utilities instead to convert it into an object.

Comment (epenet) [reply spr.py:82]: Yeah I know - I am just comparing your script with the traces you sent me previously and your parsing/splitting makes it hard to reconcile.
Maybe you need to adjust your commands to use `renault-api --log --json http get/post`
This should create log files (easier to share) and return json (easier to parse)

Comment (FRSTR) [reply spr.py:82]: Ok, so then I can just describe what is being parsed then:
1. First step is GET res-state endpoint. Again not sure what it does, maybe resets the flow, maybe checks that engine is not started already. We don't parse there anything.
2. At this step we start SRP auth. We generate so called "A" for srp based on SRP username and password, and send "A" in POST srp-sets. The length of this thing is 512 and servers checks this fact. For the POST to srp-sets, the server replies with the Notification ID, where we will then find the server result. So we need to parse the "id" in the bodyof POST response.
3. Now we GET kmr notification endpoint with the id we received above. If the body response has Salt and LoginB, we try to parse them, if not - try the same GET kmr request again after a delay. Those Salt and LoginB are the Server's response for SRP, these are used to calculate the final SRP key (M in the code, but also called "srp" in the following http request)
4. When we calculated M we can send it in the POST to engine-start. The request body has the command and M value inside the "srp" field. This is where the server can completely verify srp auth. Just like at the step 2, we don't get the server response immediately, but instead we get the "id" of the notification to check.
5. We request the notification with that new id and see the response if it fails or not.

Comment (epenet) [reply spr.py:82]: I was able to work out all that, but it is still much harder to read than if you used json :)

Comment (FRSTR) [reply spr.py:83]: I mean, come on! I basicaly googled every other line to write in python, don't make me do any of that :)

File: spr.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (FRSTR) [reply spr.py]: the number of notification responses could be different, sometimes you get the list of all jsons at once, sometimes it takes 3 notifications (especially if you don't do delays) to provide SRP_SALT_REQUEST. In the traces I had from the app, notifications with the result after step 5 took 6 notifications to provide the result.

[the remark this answers is not in the document]
Comment (FRSTR) [reply spr.py]: btw, res-state is just the state of the engine from what I see now:
1. this returned when the engine is started and running
{"data":{"type":"ResState","id":"X7LABCD1234567890","attributes":{"details":"Running","code":"42"}}}
2. this returned when the engine is stopped
{'data': {'type': 'ResState', 'id': 'X7LABCD1234567890', 'attributes': {'details': 'Stopped, ready for RES', 'code': '10'}}}

[the remark this answers is not in the document]
Comment (epenet) [reply spr.py]: That's great - do you want to add these to RenaultVehicle class (ie. replicate the work from Lock-Status)?

[the remark this answers is not in the document]
Comment (FRSTR) [reply spr.py]: Yeah, I can do that tomorrow.

**Comment by jumpjack** [comment]
I try writing down the procedure, is this correct? But which is the raw url for "GET kmr notification"? And is [this ](https://github.com/MrAspect/SRP6-Secure-Remote-Password)the javascript equivalent for SPR authentication?

```

">>>" = sent
"<<<" = received


>>>
/commerce/v1/accounts/AAAAAAAAAAAAAA/kamereon/kca/car-adapter/v1/cars/VVVVVVVVVVVVVV/actions/srp-sets?country=COUNTRY


{
  "data":
  {
    "type":"SrpSets",
    "attributes":
    {
      "a" : A.hex().upper(),  // A = second result from usr.start_authentication() or from SRP.generate_a() in [this repo](https://github.com/mitchellrj/kamereon-python/blob/146904802301aa0b0008e2bdb3a88ed10ff50acf/kamereon/kamereon.py)
      "i" : UserID            // = personId
    }
  }
}

<<< Notification ID
>>> GET kmr notification endpoint (?)
<<< Salt and LoginB (used to calculate the final SRP key "M")
>>> Send request, with "M" in body, inside the "srp" field
<<< "id" of the notification to check
>>> Request the notification with that new id and see the response if it fails or not


```

**Comment by FRSTR** [comment]
The URL for notifications is https://api-wired-prod-1-euw1.wrd-aws.com/commerce/v1/persons/<UserID>/notifications/kmr?&notificationId=<NotificationID>

And from what I remember you can skip the last part after "kmr" and it will give you all your notifications in a single response.

Regarding SRP implementation, there are different ones exist, and several exist for python. And it is known that the SRP protocol spec is not 100% strict for the implementation details, plus there were different versions and flavors on the market. This was never a big industry standard, though Apple used it with its own implementation and some other companies. So in the end, different SRP implementations will not work with each other just because they have minor differences in the way they calculate crypto payloads. I actually stumbled upon a github community that tried to fix this issue and list compatible ones, but they did not went too far and still had some questions on the spec interpretation.

I don't think I can comment on your code, as I already forgot some of the details. But you can refer to my SRP sample - that's best of my previous understanding anyway.

**Comment by jumpjack** [comment]
I don't understand if this pull request is currently working and if SRP authentication is requested also for lock-status or just for engine start.

**Comment by epenet** [comment]
SRP is only required for remote engine start.

Lock-status is working fine already, but is not supported by all vehicles.

**Comment by jumpjack** [comment]
I don't know how to install/use a not-accpeted PR.

**Comment by jumpjack** [comment]
Which values did you test as "pin"?

**Comment by jumpjack** [comment]
update: apparently in iOS app it is now possible to control horn and lights (but I have an android phone...)

**Pull request closed without merging**